packages/doenetml-worker-rust/lib-doenetml-core/src/core/dependency_creation.rs

```diff
@@ -254,23 +259,114 @@ impl Core {
                     }
                 }
             }
-            // Create a dependency from all children with matching `component_type`
-            DataQuery::ChildElementRefs { component_type } => {
+            DataQuery::FilteredChildren {
+                filters,
+                include_if_missing_profile,
+            } => {
                 let children_virtual_node = self
                     .structure_graph
                     .get_component_children_virtual_node(GraphNode::Component(
                         prop_pointer.component_idx,
                     ));
 
-                for node in self
+                // create vector of content children so that don't borrow core in loop
+                // and can make a mutable borrow of core to create a virtual node
+                let content_children = self
                     .structure_graph
                     .get_content_children(children_virtual_node)
-                {
-                    if let GraphNode::Component(component_idx) = node {
-                        let component = &self.components[component_idx];
-                        if component.get_component_type() == component_type {
-                            self.dependency_graph.add_edge(query_node, node);
-                            linked_nodes.push(node);
+                    .collect::<Vec<_>>();
+
+                // We will exclude children that are not components if there is a component type filter
+                // that restricts to a particular component type
+                let exclude_non_components = filters.iter().any(|filter| {
+                    if let DataQueryFilter::ComponentType(component_type_filter) = filter {
+                        component_type_filter.comparison == DataQueryFilterComparison::Equal
+                    } else {
+                        false
+                    }
+                });
+
+                for node in content_children {
+                    match node {
+                        GraphNode::Component(component_idx) => {
+                            let component = &self.components[component_idx];
+
+                            let mut exclude_via_component_type = false;
+
+                            let matching_props = filters
+                                .iter()
+                                .filter_map(|filter| match filter {
+                                    DataQueryFilter::PropProfile(profile_filter) => component
+                                        .provided_profiles()
+                                        .into_iter()
+                                        .find_map(|(prop_profile, local_prop_idx)| {
```

Comment:
This is at least the 3rd time we used this same filtering, right? We should make a utility function.

Reply:
Makes sense